spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/DefaultDataAccessStrategy.java

```diff
@@ -481,4 +526,31 @@ private SqlGenerator sql(Class<?> domainType) {
 			return it -> false;
 		}
 	}
+	@Nullable
+	private <T> Number getVersion(T instance, RelationalPersistentEntity<T> entity, ConversionService conversionService) {
+		RelationalPersistentProperty versionProperty = entity.getRequiredVersionProperty();
+		PersistentPropertyAccessor<T> propertyAccessor = entity.getPropertyAccessor(instance);
+		ConvertingPropertyAccessor<T> convertingPropertyAccessor = new ConvertingPropertyAccessor<>(propertyAccessor, conversionService);
+		return convertingPropertyAccessor.getProperty(versionProperty, Number.class);
+	}
+
+	private <T> Number getNextVersion(T instance, RelationalPersistentEntity<T> entity, ConversionService conversionService) {
+		Number version = getVersion(instance, entity, conversionService);
```

Comment:
How does this fail if we just use always `Long`? I'd have thought that it would simply work.

Reply:
Yes, if I use a `ConvertingPropertyAccessor`, it automatically converts from Long to the required type. The next push will contain a simplified version.

And no, it currently does not work with an immutable version field. I'll add a test. I see the merit in having an immutable version field, but I fail to see how the code in `AggregateChange` handles modifying an immutable ID field. It uses a `PropertyAccessor`, too, doesn't it?

Reply:
The trick is that the `PropertyAccessor` creates a new bean with the new value for id/version and everything after that uses the new bean returned by `propertyAccessor.getBean()`. See the end of the method I linked to.

Reply:
Ok, I get that. But in my case it fails even earlier with an `UnsupportedOperationException` right here: https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/mapping/model/BeanWrapper.java#L85

`BeanWrapper` seems to expect a wither method. Is the existence of a wither method a valid constraint for our use case?